pyanaconda/modules/payloads/payload/flatpak/source.py

```diff
@@ -0,0 +1,433 @@
+        """Installed size, in bytes"""
+        return int(self.labels["org.flatpak.installed-size"])
+
+
+class FlatpakSource(ABC):
+    """Base class for places where Flatpak images can be downloaded from."""
+
+    @abstractmethod
+    def calculate_size(self, refs: List[str]) -> Tuple[int, int]:
+        """Calculate the total download and installed size of the images in refs and
+        their dependencies.
+
+        :param refs: list of Flatpak references
+        :returns: download size, installed size
+        """
+
+    @abstractmethod
+    def download(self, refs: List[str], download_location: str,
+                 progress: Optional[ProgressReporter] = None) -> Optional[str]:
+        """Downloads the images referenced by refs and any dependencies.
+
+        If they are already local, or they can be installed
+        directly from the remote location, nothing will be downloaded.
+
+        Whether or not anything as been downloaded, returns
+        the specification of a sideload repository that can be used to install from
+        this source, or None if none is needed.
+
+        :param refs: list of Flatpak references
+        :param download_location: path to location for temporary downloads
+        :param progress: used to report progress of the download
+        :returns sideload location, including the transport (e.g. oci:<path>), or None
+        """
+
+    @property
+    @abstractmethod
+    def _images(self) -> List[SourceImage]:
+        """All images in the source, filtered for the current architecture."""
+        ...
+
+    def _expand_refs(self, refs: List[str]) -> List[str]:
+        """Expand the list of refs to be in full form and include any dependencies."""
+        result = []
+        for ref in refs:
+            # We don't do anything with the collection ID for now
+            _, ref = _canonicalize_flatpak_ref(ref)
+            result.append(ref)
+
+        for image in self._images:
+            if image.ref not in result:
+                continue
+
+            metadata = image.labels.get("org.flatpak.metadata")
+            if metadata is None:
+                continue
+
+            cp = ConfigParser(interpolation=None)
+            cp.read_string(metadata)
+            try:
+                runtime = cp.get('Application', 'Runtime')
+                if runtime:
+                    runtime_ref = f"runtime/{runtime}"
+                    if runtime_ref not in result:
+                        result.append(runtime_ref)
+            except (NoSectionError, KeyError):
+                pass
+
+        return result
+
+
+class StaticSourceImage(SourceImage):
+    """One image of a FlatpakStaticSource."""
+
+    def __init__(self, digest, manifest_json, config_json):
+        self.digest = digest
+        self.manifest_json = manifest_json
+        self.config_json = config_json
+
+    @property
+    def labels(self):
+        return self.config_json["config"]["Labels"]
```

Comment:
It would be great if we can just use a tool for this data grabbing. The tool could be part of the flatpak package and be inline when the metadata will change. In general, if possible we are trying to avoid having too much dependency on someones metadata in Anaconda because these could change in time and most of the time the authors don't know we are using it.
From a quick look it seems that half of the `source.py` code is about metadata parsing and reading.

This is probably for future improvement but I would like to know what you are thinking about it.

Reply:
If it is harder to decouple I guess it is fine then.

Reply:
This is only a json reader, we are not really doing magic here, just taking a json and reading values from it. Also if I understand it correctly these are standardized OCI labels which are used for OCI containers everywhere. I honestly think that we will start doing something similar to Image Mode too. 

It would be nice to have a tooling for that but I don't think it is requirement right now.

Reply:
Also given the fact that the labels are OCI, if there will be backward compatibility change (in how these values are read) Anaconda is not the biggest problem the OCI containers will be facing. The specific values which are used here and by Flatpak may change (that would also break stuff) but that is not hard to fix and tool for us would be in the same situation as our code.